topi/include/topi/nn/adaptive_pooling.h

```diff
@@ -0,0 +1,140 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \brief Pooling op constructions
+ * \file nn/adaptive_pooling.h
+ */
+#ifndef TOPI_NN_ADAPTIVE_POOLING_H_
+#define TOPI_NN_ADAPTIVE_POOLING_H_
+
+#include <string>
+#include <vector>
+#include <algorithm>
+
+#include "tvm/tvm.h"
+#include "tvm/ir_pass.h"
+#include "topi/tags.h"
+#include "topi/detail/pad_utils.h"
+#include "topi/nn.h"
+#include "topi/nn/pooling.h"
+
+namespace topi {
+namespace nn {
+using namespace tvm;
+
+inline Expr start_index(Expr index, Expr odim, Expr idim) {
+  return tvm::floor((index * idim) / odim);
+}
+
+inline Expr end_index(Expr index, Expr odim, Expr idim) {
+  return tvm::ceil((index + 1) * idim / odim);
+}
+
+/*!
+* \brief Perform adaptive pooling on height and width dimension of data.
+*
+* \param x The input tensor
+* \param output_size Vector of two ints: {output_height, output_width}
+* \param pool_type The type of pooling operator
+* \param height_axis index of the height dimension
+* \param width_axis index of the width dimension
+*
+* \return The output tensor in same layout order
+*/
+inline Tensor adaptive_pool_impl(const Tensor& x,
+                                 const Array<Expr>& output_size,
+                                 PoolType pool_type,
+                                 const size_t height_axis,
+                                 const size_t width_axis) {
+  CHECK(x->shape.size() >= 2) << "Adaptive Pooling input must >= 2-D (H, W)";
```

Comment:
Can the input shape be other shape besides 2D and 4D?

Reply:
Probably best to assume it's only 4D but I think greater than 4D is supported.